### PR TITLE
[BE]: Update CUDNN for Linux to 9.5.1.17 for 12.6 only

### PR DESCRIPTION
**Pull request by Skylion007**
* Significantly faster, better CUDNN Attention especially on Hopper (FA3 implementation?)
* Lots of bugfixes
* Better performance
* More numerically stable / fixed heuristics
* More functionality for SDPA

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by Skylion007** [comment]
@eqy Do we need to update any checks for CUDNN Attention here to take advantage of the new features, or do we just leave it as is for now and fix in a subsequent PR?

**Comment by eqy** [comment]
> @eqy Do we need to update any checks for CUDNN Attention here to take advantage of the new features, or do we just leave it as is for now and fix in a subsequent PR?

Let's do it step by step in case something breaks ;)

**Review by eqy — APPROVED** [review]

**Comment by eqy** [comment]
CC @nWEIdia

File: .ci/docker/common/install_cudnn.sh [hunk 1]
```diff
@@ -5,9 +5,9 @@ if [[ -n "${CUDNN_VERSION}" ]]; then
     mkdir tmp_cudnn
     pushd tmp_cudnn
     if [[ ${CUDA_VERSION:0:2} == "12" ]]; then
-        CUDNN_NAME="cudnn-linux-x86_64-9.1.0.70_cuda12-archive"
+        CUDNN_NAME="cudnn-linux-x86_64-9.5.0.50_cuda12-archive"
     elif [[ ${CUDA_VERSION:0:2} == "11" ]]; then
-        CUDNN_NAME="cudnn-linux-x86_64-9.1.0.70_cuda11-archive"
```
Comment (nWEIdia) [inline .ci/docker/common/install_cudnn.sh:10]: Should we switch to the earlier pattern that different CUDA using different cudnn version? i.e. just updating the cudnn version for cuda 12.4.1?

**Review by drisspg — APPROVED** [review]
Looks good, any reason why the update to accuracy?

**Review by nWEIdia — APPROVED** [review]

**Comment by Skylion007** [comment]
> Looks good, any reason why the update to accuracy?

Better heuristics / accuracy as a result of CUDNN would have caused unexpected successes otherwise

**Comment by Skylion007** [comment]
@drisspg How should I land this? I think I need the CUDNN uploaded to the S3.

**Comment by drisspg** [comment]
Just to confirm, what else is needed
1. Land: https://github.com/pytorch/builder/pull/2014 in parallel which updates windows builds?
2. And then making sure the wheels are upload to org/whl/nvidia-cudnn-cu12/ ?

@atalman 
I do see this file: https://github.com/pytorch/test-infra/blob/main/s3_management/update_dependencies.py
We just need to upload the newest version from here right: https://pypi.org/project/nvidia-cudnn-cu12/#history?

Not sure if this section is still up to date: https://github.com/pytorch/builder/blob/bcd0972459afd130a1c44b7386ae10c69cc1d30b/CUDA_UPGRADE_GUIDE.MD#upgrade-cudnn-version-only

**Comment by nWEIdia** [comment]
I realize test-infra windows AMI changes may also be needed: example https://github.com/pytorch/test-infra/pull/1523/files

**Comment by Skylion007** [comment]
> 2\. And then making sure the wheels are upload to org/whl/nvidia-cudnn-cu12/ ?

Wheels 

> Just to confirm, what else is needed
> 
> 1. Land: [Update to cudnn 9.5.0.50 builder#2014](https://github.com/pytorch/builder/pull/2014) in parallel which updates windows builds?
> 2. And then making sure the wheels are upload to org/whl/nvidia-cudnn-cu12/ ?
> 
> @atalman I do see this file: https://github.com/pytorch/test-infra/blob/main/s3_management/update_dependencies.py We just need to upload the newest version from here right: https://pypi.org/project/nvidia-cudnn-cu12/#history?
> 
> Not sure if this section is still up to date: https://github.com/pytorch/builder/blob/bcd0972459afd130a1c44b7386ae10c69cc1d30b/CUDA_UPGRADE_GUIDE.MD#upgrade-cudnn-version-only

Yeah, just need to upload the newest versions there. It would be great if we could mirror Nvidia pypi wheels more generally / automatically for a better OSS update process, btw.

**Comment by drisspg** [comment]
@Skylion007 the 2.5 release is being kicked off right now I am sure that @atalman is pretty busy this week but let me see if there is anything I can do to unblock

**Review by atalman — APPROVED** [review]
@Skylion007 could you please change this PR to only modify linux wheels ? We would need to update  and deploy the new AMI in order to apply this to Windows

**Comment by Skylion007** [comment]
I don't see cudnn_graph.so linked in the runtime path for the manywheel so I'm a bit concerned about that. Not sure why it's not included in those builds, they do seem to be install the cudnn wheel properly now.

**Comment by Skylion007** [comment]
This is concering, why is only the rnn_smoketest failing? Any idea why it can't find cudnn_graph.so @eqy?

**Comment by atalman** [comment]
@nWEIdia @eqy @drisspg we should be checking signal on this PR test: https://github.com/pytorch/pytorch/pull/138124
The failures with the builds on this original PR are expected

**Comment by atalman** [comment]
Issue was with docker images. Testing here: https://github.com/pytorch/pytorch/pull/138124

**Review by malfet — APPROVED** [review]
Sure. Also, I suggest just ignore, slowly stop testing Windows, as WSL is good enough alternative

**Comment by atalman** [comment]
@nWEIdia @ptrblck 
Ok looks like the reason for above faillure we can't load libcudnn_graph.so.9:

```
[root@d7973b136dbb lib]# ldd libcudnn_graph.so.9
ldd: warning: you do not have execution permission for `./libcudnn_graph.so.9'
./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./libcudnn_graph.so.9)
./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by ./libcudnn_graph.so.9)
./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./libcudnn_graph.so.9)
        linux-vdso.so.1 =>  (0x00007fffc694e000)
        libz.so.1 => /usr/lib64/libz.so.1 (0x00007f517033c000)
        librt.so.1 => /usr/lib64/librt.so.1 (0x00007f5170134000)
        libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007f516ff18000)
        libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007f516fd14000)
        libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f516fa0c000)
        libm.so.6 => /usr/lib64/libm.so.6 (0x00007f516f70a000)
        libgcc_s.so.1 => /usr/lib64/libgcc_s.so.1 (0x00007f516f4f4000)
        libc.so.6 => /usr/lib64/libc.so.6 (0x00007f516f126000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f5170b7b000)
```

**Comment by Skylion007** [comment]
> Ok looks like the reason for above faillure we can't load libcudnn_graph.so.9:
> 
> ```
> [root@d7973b136dbb lib]# ldd libcudnn_graph.so.9
> ldd: warning: you do not have execution permission for `./libcudnn_graph.so.9'
> ./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./libcudnn_graph.so.9)
> ./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by ./libcudnn_graph.so.9)
> ./libcudnn_graph.so.9: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./libcudnn_graph.so.9)
>         linux-vdso.so.1 =>  (0x00007fffc694e000)
>         libz.so.1 => /usr/lib64/libz.so.1 (0x00007f517033c000)
>         librt.so.1 => /usr/lib64/librt.so.1 (0x00007f5170134000)
>         libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007f516ff18000)
>         libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007f516fd14000)
>         libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f516fa0c000)
>         libm.so.6 => /usr/lib64/libm.so.6 (0x00007f516f70a000)
>         libgcc_s.so.1 => /usr/lib64/libgcc_s.so.1 (0x00007f516f4f4000)
>         libc.so.6 => /usr/lib64/libc.so.6 (0x00007f516f126000)
>         /lib64/ld-linux-x86-64.so.2 (0x00007f5170b7b000)
> ```
> 
> This can be resolved by updating glibcxx on CentOS 7 images: https://stackoverflow.com/questions/44205687/glibcxx-3-4-21-not-found-on-centos-7

@nWEIdia @eqy I didn't see this in the changelog. Is it intentional?

**Comment by atalman** [comment]
Suggestion for this is to go ahead with https://github.com/pytorch/pytorch/issues/123649

**Comment by nWEIdia** [comment]
In addition to the #123649 above that Andrey mentioned (to fix the libcudnn_graph.so error), we recommend NOT updating the cudnn version of cu11.8. The build may be successful, though the combination of cu11.8 and cudnn9.5 is not tested anywhere else and could invite problems that we do not have priority to fix. Let's just update the cuda 12 and cudnn9.5 combination. Thanks!

File: .github/scripts/generate_binary_build_matrix.py [hunk 2]
```diff
@@ -45,7 +45,7 @@
         "nvidia-cuda-nvrtc-cu11==11.8.89; platform_system == 'Linux' and platform_machine == 'x86_64' | "  # noqa: B950
         "nvidia-cuda-runtime-cu11==11.8.89; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cuda-cupti-cu11==11.8.87; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-cudnn-cu11==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
Comment (nWEIdia) [inline .github/scripts/generate_binary_build_matrix.py:48]: Let's not change this.

**Review by nWEIdia — DISMISSED** [review]
Per recommendation, let's keep cudnn 9.1.0.70 with cuda 11.8.

File: .github/scripts/generate_binary_build_matrix.py (hunk 2 above)
Comment (Skylion007) [reply .github/scripts/generate_binary_build_matrix.py:48]: Why not?

Comment (nWEIdia) [reply .github/scripts/generate_binary_build_matrix.py:48]: Oh, it is not changed. Not sure why I left this comment... Closing.

File: .ci/docker/common/install_cuda.sh [hunk 3]
```diff
@@ -3,7 +3,7 @@
 set -ex
 
 NCCL_VERSION=v2.21.5-1
-CUDNN_VERSION=9.1.0.70
```
Comment (nWEIdia) [inline .ci/docker/common/install_cuda.sh:6]: Reminder that we will not be using this variable but hard-code cudnn version to be 9.1.0.70 in install_118 step in this file.

**Comment by Skylion007** [comment]
@nWEIdia reverted the CUDNN changes

**Comment by eqy** [comment]
> Sure. Also, I suggest just ignore, slowly stop testing Windows, as WSL is good enough alternative

Are we testing WSL? just thinking about this as we might want WSL signal for things like https://github.com/huggingface/diffusers/issues/9704#issuecomment-2436482558

**Comment by Skylion007** [comment]
> > Sure. Also, I suggest just ignore, slowly stop testing Windows, as WSL is good enough alternative
> 
> Are we testing WSL? just thinking about this as we might want WSL signal for things like [huggingface/diffusers#9704 (comment)](https://github.com/huggingface/diffusers/issues/9704#issuecomment-2436482558)

We are not as of yet anywhere in our CD, I think

**Comment by ptrblck** [comment]
What's the status of this PR? 
Windows builds were updated to 9.5.0.50 in: https://github.com/pytorch/builder/pull/2014 while this PR tries to update Linux builds to 9.5.1.17 (and fails due to conflicts).

I would recommend not to diverge these builds too far as testing and verification could become confusing and painful.

CC @nWEIdia do you know if the needed base OS updates were done and could try to align all builds again, please?

**Comment by nWEIdia** [comment]
Once https://github.com/pytorch/pytorch/pull/138732 goes in, I believe we can rebase this PR and get the signals needed to merge this PR.

**Comment by atalman** [comment]
hi @Skylion007 please rebase and use cudnn 9.5.1 only for cuda 12.6

File: benchmarks/dynamo/ci_expected_accuracy/aot_eager_timm_training.csv [hunk 4]
```diff
@@ -218,7 +218,7 @@ tf_mixnet_l,pass,6
 
 
 
-tinynet_a,fail_accuracy,6
+tinynet_a,pass,6
```
Comment (Skylion007) [inline benchmarks/dynamo/ci_expected_accuracy/aot_eager_timm_training.csv:221]: Not upgrading this @nWEIdia will cause this to fail accuracy on CUDA 11 and pass on CUDA 12 though...

**Comment by Skylion007** [comment]
@nWEIdia It occured to me that if we leave CUDA 11.8 on 9.1 then some tests will have accuracy failures, or conversly some tasks will have unexpected successes on the inductor tests. Not sure how to handle that.

File: benchmarks/dynamo/ci_expected_accuracy/aot_eager_timm_training.csv (hunk 4 above)
Comment (nWEIdia) [reply benchmarks/dynamo/ci_expected_accuracy/aot_eager_timm_training.csv:221]: To clarify,   cuda 12 + 9.5.0.50 would make it pass, but cuda 11 + 9.1 would maintain as a failure, right? 

In the past, I had to deal with a similar situation where different CUDA version would yield different accuracy results: e.g. https://github.com/pytorch/pytorch/pull/127150/  but it seems to have too much maintenance burden. 

Another issue is: tinynet_a accuracy mismatch seem to be the issue when we update cudnn from v8 to v9: 
https://github.com/pytorch/pytorch/issues/127990

File: .github/scripts/generate_binary_build_matrix.py [hunk 5]
```diff
@@ -58,7 +58,7 @@
         "nvidia-cuda-nvrtc-cu12==12.4.127; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cuda-runtime-cu12==12.4.127; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cuda-cupti-cu12==12.4.127; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-cudnn-cu12==9.1.0.70; platform_system == 'Linux' and platform_machine == 'x86_64' | "
+        "nvidia-cudnn-cu12==9.5.1.17; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
Comment (atalman) [inline .github/scripts/generate_binary_build_matrix.py:61]: Hi @Skylion007 lets keep 9.1.0.70 for CUDA 12.4 builds. Only update 12.6

Comment (Skylion007) [reply .github/scripts/generate_binary_build_matrix.py]: I really disagree we should do this, but okay, done.

File: .ci/docker/common/install_cuda.sh [hunk 6]
```diff
@@ -3,7 +3,7 @@
...
+CUDNN_VERSION=9.5.1.17
```
Comment (malfet) [inline .ci/docker/common/install_cuda.sh:6]: No that it matters, but PR would have been 3 lines shorter, if global CUDNN_VERSION to be kept at 9.1, but 12.6 one updated to 9.5

@atalman , @seemethere : we need to pay attention to metadata and  make sure that poetry is still usable for cuda-12.6 if we keep different versions for different releases (I suspect right now it is not)

Comment (Skylion007) [reply .ci/docker/common/install_cuda.sh:6]: @malfet Intentional, I want those lines deleted 1 by 1 in the future.

File: .ci/docker/common/install_cudnn.sh [hunk 7]
```diff
@@ -4,6 +4,8 @@ if [[ -n "${CUDNN_VERSION}" ]]; then
     # cuDNN license: https://developer.nvidia.com/cudnn/license_agreement
     mkdir tmp_cudnn
     pushd tmp_cudnn
+    if [[ ${CUDA_VERSION:0:4} == "12.6" ]]; then
+        CUDNN_NAME="cudnn-linux-x86_64-9.5.1.17_cuda12-archive"
     if [[ ${CUDA_VERSION:0:2} == "12" ]]; then
```
Comment (nWEIdia) [inline .ci/docker/common/install_cudnn.sh:9]: if [[ ${CUDA_VERSION:0:2} == "12" ]]; then ...

it seems to me that CUDA 12.6 will evaluate to true for this one ? 

Perhaps, before that, is this missing a fi?  

Perhaps we need to change the if to elif?  

cc @Skylion007

**Review by nWEIdia — CHANGES_REQUESTED** [review]
install_cudnn.sh seems to have a if else syntax issue.

**Review by nWEIdia — APPROVED** [review]
Thanks for the update!

File: .ci/docker/common/install_cuda.sh (hunk 6 above)
Comment (atalman) [reply .ci/docker/common/install_cuda.sh:6]: @malfet Metadata for different system should be used from CUDA 12.4 - the one we publish to pypi. It should not affect poetry issue.

**Pull request closed without merging**